interfaces/tests/sync_int/sync_int.c

```diff
@@ -0,0 +1,211 @@
+}
+
+static int increment_on_odd_values(void* address)
+{
+    volatile_atomic int32_t* ptr = (volatile_atomic int32_t*)address;
+    while (interlocked_add(ptr, 0) < 98)
+    {
+        int value = interlocked_add(ptr, 0);
+        while (value % 2 != 1)
+        {
+            wait_on_address(ptr, &value, UINT32_MAX);
+            value = interlocked_add(ptr, 0);
+        }
+        interlocked_increment(ptr);
+        wake_by_address_all(ptr);
+    }
+    ThreadAPI_Exit(0);
+    return 0;
+}
+
+static int increment_on_even_values(void* address)
+{
+    volatile_atomic int32_t* ptr = (volatile_atomic int32_t*)address;
+    while(interlocked_add(ptr, 0) < 99)
+    {
+        int value = interlocked_add(ptr, 0);
+        while(value % 2 != 0)
+        {
+            wait_on_address(ptr, &value, UINT32_MAX);
+            value = interlocked_add(ptr, 0);
+        }
+        interlocked_increment(ptr);
+        wake_by_address_all(ptr);
+    }
+    ThreadAPI_Exit(0);
+    return 0;
+}
+
+static volatile_atomic int32_t create_count;
+static int increment_on_wake_up(void* address)
+{
+    volatile_atomic int32_t* ptr = (volatile_atomic int32_t*)address;
+    int32_t value = interlocked_add(ptr, 0);
+    interlocked_increment(&create_count);
+    wake_by_address_single(&create_count);
+    wait_on_address(ptr, &value, UINT32_MAX);
+    interlocked_increment(ptr);
+    ThreadAPI_Exit(0);
+    return 0;
+}
+
+
+BEGIN_TEST_SUITE(sync_int)
+
+
+TEST_SUITE_INITIALIZE(a)
+{
+    g_testByTest = TEST_MUTEX_CREATE();
+    ASSERT_IS_NOT_NULL(g_testByTest);
+}
+
+TEST_SUITE_CLEANUP(b)
+{
+
+    TEST_MUTEX_DESTROY(g_testByTest);
+}
+
+TEST_FUNCTION_INITIALIZE(c)
+{
+    if (TEST_MUTEX_ACQUIRE(g_testByTest))
+    {
+        ASSERT_FAIL("our mutex is ABANDONED. Failure in test framework");
+    }
+}
+
+TEST_FUNCTION_CLEANUP(d)
+{
+    TEST_MUTEX_RELEASE(g_testByTest);
+}
+
+/*Codes_SRS_SYNC_43_001: [ wait_on_address shall atomically compare *address and *compare_address.]*/
+/*Codes_SRS_SYNC_43_007: [ If *address is equal to *compare_address, wait_on_address shall cause the thread to sleep. ]*/
+/*Codes_SRS_SYNC_43_008: [wait_on_address shall wait indefinitely until it is woken up by a call to wake_by_address_[single/all] if timeout_ms is equal to UINT32_MAX]*/
+/*Codes_SRS_SYNC_43_003: [ wait_on_address shall wait until another thread in the same process signals at address using wake_by_address_[single/all] and return true. ]*/
+TEST_FUNCTION(two_threads_increment_alternately)
+{
+    ///arrange
+    volatile_atomic int32_t var;
+    interlocked_exchange(&var, 0);
+
+    THREAD_HANDLE thread1;
+    THREAD_HANDLE thread2;
+    ///act
+    ASSERT_ARE_EQUAL(THREADAPI_RESULT, THREADAPI_OK, ThreadAPI_Create(&thread1, increment_on_even_values, (void*)&var));
+    ASSERT_ARE_EQUAL(THREADAPI_RESULT, THREADAPI_OK, ThreadAPI_Create(&thread2, increment_on_odd_values, (void*)&var));
+
+
+
+    ///assert
+    int return1;
+    ThreadAPI_Join(thread1, &return1);
```

Comment:
>ThreadAPI_Join [](start = 4, length = 14)

I think these should also be wrapped in `ASSERT_ARE_EQUAL(THREADAPI_RESULT, THREADAPI_OK,...)`. Ok to ignore the `return1` and `return2` since you always use 0.